### PR TITLE
DOCSP-32634: Add Indexes page

### PR DESCRIPTION
**Pull request by mcmorisi**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - [DOCSP-32634](https://jira.mongodb.org/browse/DOCSP-32634)
Staging - <https://preview-mongodbmcmorisi.gatsbyjs.io/csharp/DOCSP-32634-indexes/fundamentals/indexes/>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/fundamentals/indexes.txt [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+   :end-before: end-compound-index-query
+   :dedent:
+
+For more information, see :manual:`Compound Indexes </core/index-compound>` in the Server manual.
+
+Multikey Indexes
+~~~~~~~~~~~~~~~~
+
+**Multikey indexes**  are indexes that improve performance for queries that specify a
+field with an index that contains an array value. You can define a multikey index using
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:146]: Although this is the description we use on other pages, I think it's too wordy and hard to understand. I think the description they use on the Server docs is a little better:
```suggestion
**Multikey indexes** collect and sort data from fields containing array values. You can define a multikey index using
```

File: source/fundamentals/indexes.txt [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+Clustered Indexes
+~~~~~~~~~~~~~~~~~
+
+**Clustered indexes** instruct a collection to store documents ordered by a key value.
+To create a clustered index, specify the clustered index
+option with the ``_id`` field as the key and the ``Unique`` field as
+``true`` when you create your collection.
+
+The following example creates a clustered index on the ``_id`` field in
+the ``tea.vendors`` collection:
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:181]: To be more inline with the other examples on this page, can we make this example create a new collection in the sample_mflix database instead of a new tea database? I also think it could be reworded a bit:
```suggestion
The following example creates a clustered index on the ``_id`` field while creating a new ``tea.vendors`` collection:
```

File: source/fundamentals/indexes.txt [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+    };
+
+    teaDatabase.CreateCollection("vendors", new CreateCollectionOptions<Vendor>
+    {
+        ClusteredIndex = clusteredIndexOptions
+    });
+
+To learn more, see
+:v6.0:`Clustered Indexes </reference/method/db.createCollection/#std-label-db.createCollection.clusteredIndex>` and
+:v6.0:`Clustered Collections </core/clustered-collections>`.
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:199]: Update to :manual: tags so it's not linking to an old version
```suggestion
:manual:`Clustered Indexes </reference/method/db.createCollection/#std-label-db.createCollection.clusteredIndex>` and
:manual:`Clustered Collections </core/clustered-collections>` in the Server manual.
```

File: source/includes/fundamentals/code-examples/Indexes.cs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        }
+        // Prints a message if a text index already exists with a different configuration 
+        catch (MongoCommandException e)
+        {
+            if (e.CodeName == "IndexOptionsConflict")
+            {
+                Console.WriteLine("There is an existing text index with different options");
+            }
+        }
+        // end-text-index
```
Comment (jordan-smith721) [inline source/includes/fundamentals/code-examples/Indexes.cs:110]: Is there a reason you're including the try/catch block on this example but not on the others? If it's not explicitly necessary I think we should just show the line that creates the index so that it mirrors the other examples.

File: source/fundamentals/indexes.txt [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+index.
+
+The following snippet creates a compound text index for the ``title`` and ``genre``
+fields within the ``sample_mflix.movies`` collection:
+
+.. code-block:: csharp
+
+    collection.Indexes.CreateOne("{ title : \"text\", genre: \"text\" }");
+
+For more information, see :manual:`Compound Text Index Restrictions </core/index-text/#std-label-text-index-compound>`
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:251]: I think this section moved:
```suggestion
For more information, see :manual:`Compound Text Index Restrictions </core/indexes/index-types/index-text/text-index-restrictions/#compound-text-index>`
```

File: source/fundamentals/indexes.txt [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+Geospatial Indexes
+~~~~~~~~~~~~~~~~~~
+
+MongoDB supports queries of geospatial coordinate data using **2dsphere
+indexes**. With a 2dsphere index, you can query the geospatial data for
+inclusion, intersection, and proximity.
+
+To create a 2dsphere index, you must specify a field that contains
+only **GeoJSON objects**. For more details on this type, see the MongoDB
+Server manual page on :manual:`GeoJSON objects </reference/geojson>`.
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:263]:
```suggestion
only **GeoJSON objects**. For more details on this type, see :manual:`GeoJSON objects </reference/geojson>` in the MongoDB Server manual.
```

File: source/fundamentals/indexes.txt [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+            ]
+         }
+      }
+   }
+
+The following example creates a ``2dsphere`` index on the ``location.geo`` field:
+
+.. important::
+
+   Attempting to create a geospatial index on a field that is covered by a geospatial index results in an error.
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:295]: I think this is more clear (if I got the meaning correct)
```suggestion
   Attempting to create a geospatial index on a field that is already covered by a geospatial index results in an error.
```

File: source/fundamentals/indexes.txt [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+
+.. literalinclude:: /includes/fundamentals/code-examples/Indexes.cs
+   :language: csharp
+   :start-after: begin-geospatial-query
+   :end-before: end-geospatial-query
+   :dedent:
+
+MongoDB also supports ``2d`` indexes for calculating distances on a Euclidean plane and
+for working with the "legacy coordinate pairs" syntax used in MongoDB 2.2 and earlier.
+To learn more, see :manual:`Geospatial Queries </geospatial-queries>`.
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:313]:
```suggestion
To learn more, see :manual:`Geospatial Queries </geospatial-queries>` in the Server manual.
```

File: source/fundamentals/indexes.txt [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+
+.. note::
+
+   The Atlas Search Index management methods run asynchronously. The
+   driver methods can return before confirming that they ran
+   successfully. To determine the current status of the indexes, call the
+   `IMongoSearchIndexManager.List() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.Search.IMongoSearchIndexManager.List.html>`__ method.
+
+The following sections contain links to tutorials with code examples that demonstrate
+how to create and interact with Atlas Search indexes.
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:382]: S: I think "with code examples" is implied here
```suggestion
The following sections contain links to tutorials that demonstrate
how to create and interact with Atlas Search indexes.
```

**Review by jordan-smith721 — CHANGES_REQUESTED** [review]
Nice work! A few suggestions and link fixes

File: source/includes/fundamentals/code-examples/Indexes.cs (hunk 4 above)
Comment (mcmorisi) [reply source/includes/fundamentals/code-examples/Indexes.cs:110]: Can update this one and the geospatial examples to only include the index creation line.

**Review by jordan-smith721 — APPROVED** [review]
LGTM!

File: source/fundamentals/indexes.txt [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
++++++++++++++++++++
+
+To learn how to delete an Atlas Search index using the {+driver-short+}, see
+:atlas:`Delete an Atlas Search Index </atlas-search/delete-index/>` in the Atlas manual
+and select :guilabel:`C#` from the language dropdown.
+
+List Indexes
+------------
+
+You can use the `IMongoIndexManager<TDocument>.List() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IMongoIndexManager-1.List.html>`__
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:414]: I'm not sure we need to list the interface here, imo it just makes things less clear since you don't actually call it like that in the code.

```suggestion
You can use the `List() <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.IMongoIndexManager-1.List.html>`__
```

File: source/fundamentals/indexes.txt [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+method to retrieve a list of indexes in your collection.
+
+The following example uses the ``IMongoIndexManager<TDocument>.List()`` method to list
```
Comment (jordan-smith721) [inline source/fundamentals/indexes.txt:417]:
```suggestion
The following example uses the ``List()`` method to list
```

File: source/includes/fundamentals/code-examples/Indexes.cs [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        UniqueIndex(theaterCollection);
+        WildcardIndex(theaterCollection);
+    }
+
+    private static void SingleIndex(IMongoCollection<Movie> collection)
+    {
+        Console.WriteLine("single index");
+
+        // begin-single-index
+        collection.Indexes.CreateOne("{ title : 1 }");
```
Comment (JamesKovacs) [inline source/includes/fundamentals/code-examples/Indexes.cs:37]: As noted above, I would steer users to the strongly typed `Builders<T>.IndexKeys` builder for defining indexes.

File: source/fundamentals/indexes.txt [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+   :start-after: begin-wildcard-index
+   :end-before: end-wildcard-index
+   :dedent:
+
+For more information, see the :manual:`Wildcard Indexes</core/index-wildcard>` page in the
+Server manual.
+
+.. _search-indexes:
+
+Atlas Search Indexes
```
Comment (JamesKovacs) [inline source/fundamentals/indexes.txt:362]: Given the confusion between `text` and Atlas Search indexes, I would suggest moving the two side-by-side for easier documentation and comparison. If you're following a standard cross-driver ordering, then it's fine to leave as-is.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Overall looking great. Biggest recommendation is to use the strongly-typed `Builders<T>.IndexKeys` builder rather than specifying indexes as `BsonDocument` or JSON strings.

File: source/includes/fundamentals/code-examples/Indexes.cs [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+            options);
+        // end-unique-index
+    }
+
+    private static void WildcardIndex(IMongoCollection<Theater> collection)
+    {
+        Console.WriteLine("wildcard index");
+
+        // begin-wildcard-index
+        collection.Indexes.CreateOne(Builders<Theater>.IndexKeys.Ascending("location.$**"));
```
Comment (mcmorisi) [inline source/includes/fundamentals/code-examples/Indexes.cs:175]: @JamesKovacs Wanted to specifically draw your attention here – is this the preferred way to implement an ascending wildcard index using the Builders model?

Comment (JamesKovacs) [reply source/includes/fundamentals/code-examples/Indexes.cs:175]: We have a `Wildcard` method on `IndexKeys`.

```
coll.Indexes.CreateOne(
    new CreateIndexModel<Theater>(Builders<Review>.IndexKeys.Wildcard(x => x.Location))
);
```

File: source/includes/fundamentals/code-examples/Indexes.cs [hunk 15]
```diff
@@ -34,7 +34,7 @@ private static void SingleIndex(IMongoCollection<Movie> collection)
         Console.WriteLine("single index");
 
         // begin-single-index
-        collection.Indexes.CreateOne("{ title : 1 }");
+        collection.Indexes.CreateOne(Builders<Movie>.IndexKeys.Ascending("title"));
```
[the remark this answers is not in the document]
Comment (JamesKovacs) [reply source/includes/fundamentals/code-examples/Indexes.cs:37]: Apologies as I should have caught this in my first review. This overload of `CreateOne` is deprecated.

```
Method 'MongoDB.Driver.IMongoIndexManager<TDocument>.CreateOne(IndexKeysDefinition<TDocument>, CreateIndexOptions, CancellationToken)' is obsolete: 'Use CreateOne with a CreateIndexModel instead.'
```

This applies to all `CreateOne` and `CreateMany` calls below.

```
var indexModel = new CreateIndexModel<Movie>(Builders<Movie>.IndexKeys.Ascending(x => x.Title));
coll.Indexes.CreateOne(indexModel);
```

This overload allows you to pass in an `CreateIndexOptions` on a per-index basis.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
A few more changes...

File: source/includes/fundamentals/code-examples/Indexes.cs (hunk 14 above)
Comment (mcmorisi) [reply source/includes/fundamentals/code-examples/Indexes.cs:175]: In this case, would the `x.Location` wildcard be defined in the `Theater` Poco as something like this?

```
        [BsonElement("location.$**")]
        public string LocationGeo { get; set; }
```

Comment (JamesKovacs) [reply source/includes/fundamentals/code-examples/Indexes.cs:175]: `[BsonElement]` specifies the field name in the database and is unrelated to wildcard indexes. You would still specify this field as `location.geo`.

**Review by JamesKovacs — APPROVED** [review]
LGTM
